### PR TITLE
New build environment, and CI test environments

### PR DESCRIPTION
**Pull request by mroth**
An experimental branch for us to mess with us using the scala-builder Docker
image and rationalizing a proposed automated release process.

 - A new Dockerfile, based on `openlaw/scala-builder`, that builds the compiled
   environment for ol-core  only (which in this project, is used for running tests or
   making a release). This image builds the dependencies first so that they can
   be cached.
 - A `/scripts` folder containing scripts that are responsible for running any
   CI process. Most of these are currently just wrappers around sbt commands,
   but enable us to more easily see what is possible, define requirements, and
   potentially move some steps out of sbt control in the future if needed.
 - Experimental test, coverage, and release stage integration for the following hosted CI environments:
   - CircleCI (build cached on Docker Hub), see `.semaphore` dir
   - SemaphoreCI (build cached in AWS EU region), see `.circleci` dir
   - Github Actions  (build cached on Docker Hub), see `.github` dir

All three CI environments will do a layer cached build, run tests and coverage reports in parallel with failFast reporting (coverage report is inclusive of tests but takes 2.5x longer to run, so we prefer faster failure).

Note from this PR the following were excluded:
 - JenkinsX (too unstable, config archived in builder-jenkinsx branch)
 - Google Cloud Build (requires $ and will build if any Dockerfile present, so in different repo for now to avoid auto-triggering while we test).

While all are configured, to avoid collisions, only one will do the actual pushing of coverage reports to Codacy and releases to BinTray. This is controlled via env var flags.

In the future, we will pick one and optimize (including likely upgrading machine class for faster builders), but for now the goal is just to get the automated flow working with the new images, and we can have real world comparison to discuss.

Since we are making major changes to our general develop/branching/release process, the goal is to get this integrated with the release flag *disabled*, and do manual releases using the new scripts for a bit. However, all the CI environments should be triggered in the background, and will do everything up to but excluding the final push to BinTray, giving us the flexibility to see how it is operating on real world data, so we can debug and tune before making it the default, and getting us familiar with the new branch model.

Docs related to the current manual release process are in draft here: https://github.com/openlawteam/openlaw-core/pull/123

### CODE REVIEW

**Comment by mroth** [comment]
Flow chart (incomplete) from discussion:
ol-core-flow

**Comment by mroth** [comment]
We have rebased against latest master, now investigating why build is breaking with new changes.

File: .circleci/config.yml [hunk 1]
```diff
@@ -1,51 +1,92 @@
-# Scala CircleCI 2.0 configuration file
-#
-# Check https://circleci.com/docs/2.0/sample-config/ for more details
-#
-version: 2
-jobs:
-  build:
+version: 2.1
+# https://circleci.com/docs/2.0/reusing-config/#authoring-reusable-executors
```
Comment (outkaj) [inline .circleci/config.yml]: maybe we should explain why we need to use the reusable executor?

File: .circleci/config.yml [hunk 2]
```diff
@@ -1,51 +1,92 @@
...
+      - release:
+          context: BintrayCredentials
+          requires:
+            - test
+          filters:
+            tags:
+              only: /^v.*/
+            branches:
+              ignore: /.*/
+          # sadly circle-ci only ORs, need to handle logic in release script itself?
```
Comment (outkaj) [inline .circleci/config.yml]: what does "OR" mean here?

File: .github/main.workflow [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+  runs = ["sh", "-c", "BRANCH=${GITHUB_REF##*/} ID=gh ci/build.sh"]
+  needs = ["Docker Registry Login"]
+  env = {
+    PUSH_CACHE = "0"
+  }
+}
+
+action "Caching Build" {
+  uses = "actions/docker/cli@master"
+  runs = ["sh", "-c", "BRANCH=${GITHUB_REF##*/} ID=gh ci/build.sh"]
```
Comment (outkaj) [inline .github/main.workflow]: should we get rid of the part after `REF` or is this saved here to be uncommented later? (same with the one above?)

File: Dockerfile [hunk 4]
```diff
@@ -1,67 +1,21 @@
...
-ARG RELEASE
-ENV RELEASE=${RELEASE}
-
-RUN if [ "$RELEASE" = "true" ]; then sbt 'release with-defaults'; fi
+FROM openlaw/scala-builder:node
+
+# install plugins
+COPY project ./project
+# copy .sbtopts for project-specific JVM settings (we may want to use these across
+# projects, or get rid of entirely - TBD
```
Comment (outkaj) [inline Dockerfile]: looks like I left off a close parenthesis here, sorry for that

File: README.md [hunk 5]
```diff
@@ -36,3 +37,5 @@ distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
+
+Test
```
Comment (outkaj) [inline README.md]: should we add instructions for running the tests under here?

File: project/plugins.sbt [hunk 6]
```diff
@@ -1,21 +1,24 @@
...
+addSbtPlugin("org.scala-js" % "sbt-scalajs" % "0.6.27")
+addSbtPlugin("org.portable-scala" % "sbt-scalajs-crossproject" % "0.6.0")
 
-addSbtPlugin("org.wartremover" % "sbt-wartremover" % "2.4.1")
+/* Release process */
+addSbtPlugin("com.typesafe.sbt" % "sbt-git" % "1.0.0")
+addSbtPlugin("org.foundweekends" % "sbt-bintray" % "0.5.4")
+addSbtPlugin("com.github.gseitz" % "sbt-release" % "1.0.11")
 
+/* Code coverage */
```
Comment (outkaj) [inline project/plugins.sbt]: should we add a TODO - to potentially get rid of this if we end up not using Codacy in future?

File: .github/main.workflow (hunk 3 above)
Comment (mroth) [reply .github/main.workflow]: Weirdly this is a bug in GitHub web interface's syntax highlighting.  Those hashmarks are part of a command transforming the variable to strip out everything prior to the the final slash, in order to extract just the branch name. (Isn't bash so clear and easy to read? Ugh.)

File: README.md (hunk 5 above)
Comment (mroth) [reply README.md]: I think this was just a dangling text that has been there for a while.  I'll just remove it for now.  Good catch.

File: Dockerfile (hunk 4 above)
Comment (mroth) [reply Dockerfile]: On it!

**Pull request merged**